### PR TITLE
Keep numpy sparsemax output in the input dtype

### PR DESCRIPTION
**Pull request by SEPURI-SAI-KRISHNA**
## Description

Follow-up to #23469, which fixed the sparsemax threshold. The numpy backend
still promotes its output away from the dtype of the input.

`ops.sparsemax` returns `float64` on the numpy backend for `float16`,
`float32` and `bfloat16` inputs. The jax, tensorflow and torch backends all
return the input dtype, so a model running in `float32` silently gets
`float64` activations out of this one op on this one backend.

```python
import numpy as np
from keras import ops

x = np.random.uniform(size=(2, 5)).astype("float32")
print(ops.sparsemax(x).dtype)
# numpy backend      -> float64   # wrong
# jax/tf/torch       -> float32   # correct
```

Two independent causes, both in `keras/src/backend/numpy/nn.py`:

| expression on master | float16 | float32 | bfloat16 |
| --- | --- | --- | --- |
| `-1.0 * np.sort(-1.0 * logits, axis=axis)` | ok | ok | **float32** |
| `(logits_cumsum - 1) / r` | **float64** | **float64** | **float64** |
| `np.where(support, logits_sorted, 0.0)` | ok | ok | **float64** |
| `(np.sum(...) - 1) / k` | **float64** | **float64** | **float64** |
| `np.maximum(logits - tau, 0.0)` | ok | ok | **float32** |

`r` holds the ranks `1..n` and `k` the support sizes. Both come out of
`np.arange` / `np.sum` as `int64`, and dividing a float array by an integer
array promotes to `float64` whatever the float width — that accounts for
every dtype. Separately, the Python float constants are weak under NEP 50 for
the numpy float types but not for `ml_dtypes.bfloat16`, which is why
`bfloat16` is the only dtype the constants affect.

### Fix

Materialize the ranks, the counts and the constants in the dtype of the
input, and use unary negation for the descending sort so it does not go
through a Python float multiply. The algorithm is unchanged.

### Tests

`NNOpsCorrectnessTest::test_sparsemax_dtype` is new. It asserts the output
dtype follows the input for `float16`, `float32` and `bfloat16`. On master it
fails on the numpy backend (`float64` != `float16`) and passes on the other
three, which is the inconsistency this closes.

I also added a deterministic multi-support case to the existing
`NNOpsCorrectnessTest::test_sparsemax`, which only covered
`[-0.5, 0, 1, 2, 3]`. That input has a support of size one — the shape of
input that hid the threshold bug fixed in #23469, since the ops-level test
would still have passed against the old threshold. #23469 added multi-support
coverage to `activations_test.py` but not here. `[0.0, 0.5, 1.0] ->
[0.0, 0.25, 0.75]` is exactly representable in binary floating point, so it
needs no tolerance. This case passes on master today; it is coverage against
reintroduction, not a second bug fix.

Full runs of `keras/src/ops/nn_test.py` and
`keras/src/activations/activations_test.py` on every backend I can run
locally:

| backend | result |
| --- | --- |
| numpy | 449 passed, 27 skipped |
| jax | 459 passed, 17 skipped |
| tensorflow | 462 passed, 14 skipped |
| torch | 448 passed, 28 skipped |

I do not have an OpenVINO runtime available, but this PR does not touch the
OpenVINO backend. `ruff` and `ruff-format` are clean at the pinned v0.9.2.

## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

## AI assistance disclosure

Per the AI-Assisted Contribution Policy: I used an AI coding assistant to help
investigate this dtype promotion and to draft the patch, the tests and the
test runs reported above. I have reviewed the change, I understand why
dividing a float array by an `int64` array promotes to `float64` and why
`ml_dtypes.bfloat16` does not get NEP 50 weak-scalar treatment from the Python
float constants, and I take responsibility for the code submitted here.
Review responses will be written by me.

### CODE REVIEW

File: keras/src/backend/numpy/nn.py [hunk 1]
```diff
@@ -220,18 +220,23 @@ def log_softmax(x, axis=-1):
...
+    r = r.reshape(r_shape).astype(logits.dtype)
+    support = logits_sorted - (logits_cumsum - 1) / r > zero
     # Find the threshold
-    k = np.sum(support, axis=axis, keepdims=True)
-    logits_sorted_safe = np.where(support, logits_sorted, 0.0)
+    k = np.sum(support, axis=axis, keepdims=True).astype(logits.dtype)
+    logits_sorted_safe = np.where(support, logits_sorted, zero)
     tau = (np.sum(logits_sorted_safe, axis=axis, keepdims=True) - 1) / k
-    output = np.maximum(logits - tau, 0.0)
+    output = np.maximum(logits - tau, zero)
```
[the remark this answers is not in the document]
Comment (SEPURI-SAI-KRISHNA) [reply keras/src/backend/numpy/nn.py:239]: Thanks, I measured this before writing the patch, and the integer literal is already safe. It is the *float* literals that promote `bfloat16`, which is why this PR replaces those and deliberately leaves `- 1` alone:

```
numpy 2.5.2 | ml_dtypes 0.5.4

bfloat16 - 1                     -> bfloat16
bfloat16 - 1.0                   -> float32
bfloat16 - np.array(1.0, dtype)  -> bfloat16
```

And the two expressions this comment points at, with a `bfloat16` input:

```
(logits_cumsum - 1) / r        -> bfloat16
(np.sum(...) - 1) / k          -> bfloat16
```

Both already stay in the input dtype, so `one` would not change the result. The constants the PR does materialize are the ones in `np.where(support, logits_sorted, 0.0)` and `np.maximum(logits - tau, 0.0)`, where the Python
float does promote `bfloat16`.

On the "may not hold in all environments or NumPy versions" point: rather than carry a constant that is not needed on any version I can test, `test_sparsemax_dtype` in this PR asserts the output dtype for `bfloat16` directly. If some supported NumPy / ml_dtypes combination did promote through the integer literal, that test fails in CI rather than the upcast slipping through silently. I would rather the guard live in the test.

File: keras/src/ops/nn_test.py [hunk 2]
```diff
@@ -1685,6 +1685,21 @@ def test_sparsemax(self):
...
+        self.assertAllClose(knn.sparsemax(x), [0.0, 0.25, 0.75])
+
+    def test_sparsemax_dtype(self):
+        # The ranks and counts inside the implementation are integers and the
+        # constants are Python floats. Neither may promote the result away
+        # from the dtype of the input.
+        x = np.random.uniform(size=(2, 5)).astype("float32")
+        for dtype in ("float16", "float32", "bfloat16"):
+            output = knn.sparsemax(ops.cast(x, dtype))
+            self.assertEqual(standardize_dtype(output.dtype), dtype)
```
Comment (hertschuh) [inline keras/src/ops/nn_test.py]: Please look at the `NNOpsDtypeTest` test class below and add a test in this class following the same pattern.

**Review by hertschuh — CHANGES_REQUESTED** [review]
Thanks for looking into this!

File: keras/src/ops/nn_test.py (hunk 2 above)
Comment (SEPURI-SAI-KRISHNA) [reply keras/src/ops/nn_test.py]: Moved it. `test_sparsemax_dtype` is gone from `NNOpsCorrectnessTest`, and `NNOpsDtypeTest` now has a `test_sparsemax` parameterized over `FLOAT_DTYPES` that checks both the eager call and the symbolic call, next to `test_sparse_sigmoid`.

One deviation from the surrounding tests: `jax.nn` has no `sparsemax`, so there is no `jnn.*` reference to derive `expected_dtype` from. I assert the dtype directly, the way `test_ctc_loss` and `test_ctc_decode` in the same class do. Happy to change it if you'd prefer something else there.

All three parameters fail on master (`float64` instead of the input dtype) and pass with the fix.

**Review by hertschuh — APPROVED** [review]
Thanks!
